### ISSUE TITLE
API status: ambiguous state

### ISSUE THREAD

**Issue by akaJes** [isomorphic-git/isomorphic-git#358]
ambiguous state of the file status https://isomorphic-git.org/docs/en/status
when i have the same file:

-in staged mode - it is modified
-in unstaged mode - it is modified or removed

and other cases with (added, removed, modified, renamed...)

**Comment by wmhilton** [comment]
What do you think of the `statusMatrix` command coming in #355? That should make it possible to disambiguate all possible states, and not rely on english words and asterisks.

**Comment by akaJes** [comment]
this will be more useful than `status`
> all possible states

except the `rename` state, also how about `git mv` command?

and `git.statusMatrix` looks more complicated,
for me personaly, could be better like cumlative function, from which you can decide everything
```
const status = ({fs, dir, gitdir, filepath}) => {
...
return {STAGEfilepath, HEADsha1, STAGEsha1, WORKDIRsha1}
}
```

**Comment by wmhilton** [comment]
> except the rename state

that (it turns out) is much harder to calculate. I think git uses an optimized file content similarity search to figure that out on the fly. I haven't even seen if it caches that information anywhere.

> return {STAGEfilepath, HEADsha1, STAGEsha1, WORKDIRsha1}
I would, except a lot of the time you don't **need** the WORKDIRsha1. For example, if you are adding a bunch of files, all that matters is "it's new!". Calculating the SHA sums of potentially dozens of files when all you needed was to know they are "new" would slow down the status command a ton.

**Comment by akaJes** [comment]
how to calculate the case when the file in **stage** is modified and in **workdir** also ??
HEADsha1 && HEADsha1 != STAGEsha1 && STAGEsha1
and 
STAGEsha1 && STAGEsha1 != WORKDIRsha1 && WORKDIRsha1
without sha1 it is impossible!
about a lot of files and slow down it is true - for this should be used an option to calculate or no(check only if file exists)

**Comment by akaJes** [comment]
The native git uses difflib to compare whole HEAD tree to workdir https://github.com/git/git/blob/master/wt-status.c#L727

**Comment by wmhilton** [comment]
> how to calculate the case when the file in stage is modified and in workdir also ??

yes, which is why I do calculate the SHA1 in `statusMatrix` when absolutely necessary. But if the fs.stat data shows the file hasn't been changed since the last time it was updated in the `.git/index`, etc, etc.

I can add a flag to `statusMatrix` that makes it return the absolute value of the SHA1. And that could be useful when comparing say, 10 commits. But I think as a "git status" replacement, you don't care about the absolute value of the SHA1, only how the SHAs compare to each other. Which is why I return relative values instead of absolute ones.

**Comment by wmhilton** [comment]
Well, if you really want to dive into it, here's the code that determines whether a file has been renamed:
https://github.com/git/git/blob/master/diffcore-rename.c

**Comment by akaJes** [comment]
did you measure the performance of native node sha1 vs browser sha.js ?
and would use something like universal sha? this will good option(sha1 of **workingdir** for statusMatrix)
```
require('detect-node') && require('crypto').createHash('sha1').update(buffer).digest('hex') || new Hash().update(buffer).digest('hex')
```

**Comment by mojavelinux** [comment]
I think I found a bug in the statusMatrix, but I wanted to discuss it here first to be sure before filing an issue.

Consider the following code:

```js
const fs = require('fs-extra')
const git = require('isomorphic-git')
const path = require('path')

git.plugins.set('fs', fs)

;(async () => {
  const dir = 'status-matrix-bug'
  const repo = { dir }
  const author = { name: 'Git', email: 'git@example.org' }
  await git.init(repo)
    .then(() => git.commit({ ...repo, author, message: 'init' }))
    .then(() => fs.writeFile(path.join(dir, 'README.adoc'), '= README', 'utf-8'))
    .then(() => git.add({ ...repo, filepath: 'README.adoc' }))
    .then(() => git.commit({ ...repo, author, message: 'change README' }))
    .then(() => git.branch({ ...repo, ref: 'branch-a' }))
    .then(() => git.checkout({ ...repo, ref: 'branch-a' }))
    .then(() => fs.writeFile(path.join(dir, 'README.adoc'), '= README (branch-a)', 'utf-8'))
    // statusMatrix correctly reports README.adoc as 1, 2, 1
    .then(() => git.statusMatrix(repo).then((status) => console.dir(status)))
    .then(() => git.add({ ...repo, filepath: 'README.adoc' }))
    .then(() => git.commit({ ...repo, author, message: 'change README' }))
    .then(() => git.branch({ ...repo, ref: 'branch-b' }))
    .then(() => git.checkout({ ...repo, ref: 'branch-b' }))
    .then(() => fs.writeFile(path.join(dir, 'README.adoc'), '= README (branch-b)', 'utf-8'))
    // statusMatrix incorrectly reports README.adoc as 1, 1, 1
    .then(() => git.statusMatrix(repo).then((status) => console.dir(status)))
    .then(() => git.add({ ...repo, filepath: 'README.adoc' }))
    .then(() => git.commit({ ...repo, author, message: 'change README' }))
})()
```

The status for README.adoc on the second branch is reported as 1, 1, 1 instead of the expected 1, 2, 1. For some reason, it's not considering the file in the workdir as modified, even though it is. (If I stage the file, then it is recognized as modified).

**Comment by wmhilton** [comment]
😭 No discussion needed here. I'll open an issue.
